### PR TITLE
Added a document about CORS Service.

### PR DESCRIPTION
**Pull request by tmdgus0118**
Issues #1523

### CODE REVIEW

File: site/src/sphinx/server-cors-service.rst [hunk 1]
```diff
@@ -0,0 +1,107 @@
+.. _CORSWiki: https://en.wikipedia.org/wiki/Cross-origin_resource_sharing
+.. _server-cors-service:
+
+Configure CORS service
+========================
+
+Armeria provides a way to configure Cross-origin resource sharing(CORS) policy for specified origins or
+any origin via :api:`CorsServiceBuilder`. For more information about CORS, visit CORSWiki_.
```
Comment (trustin) [inline site/src/sphinx/server-cors-service.rst]: How about:

   Visit the `Wikipedia's CORS <https://en.wikipedia...>`_ page for more information.

File: site/src/sphinx/server-cors-service.rst [hunk 2]
```diff
@@ -0,0 +1,107 @@
+.. _CORSWiki: https://en.wikipedia.org/wiki/Cross-origin_resource_sharing
+.. _server-cors-service:
+
+Configure CORS service
+========================
+
+Armeria provides a way to configure Cross-origin resource sharing(CORS) policy for specified origins or
```
Comment (trustin) [inline site/src/sphinx/server-cors-service.rst]: Missing space between `sharing` and `(CORS)`

File: site/src/sphinx/server-cors-service.rst [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+any origin via :api:`CorsServiceBuilder`. For more information about CORS, visit CORSWiki_.
+
+
+Allowing any origin
+-------------------
+To configure CORS Service allowing any origin (*), use ``CorsServiceBuilder.forAnyOrigin()``
+
+.. code-block:: java
+
+    HttpService myService = new AbstractHttpService() {
```
Comment (trustin) [inline site/src/sphinx/server-cors-service.rst]: Could use a lambda expression: `myService = (ctx, req) -> ...;`

File: site/src/sphinx/server-cors-service.rst [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+                              .allowRequestHeaders(HttpHeaderNames.of("allow_request_header"))
+                              .exposeHeaders(HttpHeaderNames.of("expose_header_1"),
+                                             HttpHeaderNames.of("expose_header_2"))
+                              .preflightResponseHeader("x-preflight-cors", "Hello CORS")
+                              .newDecorator()));
+
+
+
+
+Allowing the specified origins
```
Comment (trustin) [inline site/src/sphinx/server-cors-service.rst]: Allowing specific origins ?

File: site/src/sphinx/server-cors-service.rst [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+------------------------------
+To configure CORS Service allowing the specified origins, use ``CorsServiceBuilder.forOrigins`` or
```
Comment (trustin) [inline site/src/sphinx/server-cors-service.rst]: - the specified origins -> specific origins ?
- `forOrigins` -> `forOrigins()`
- `forOrigin` -> `forOrigin()`

File: site/src/sphinx/server-cors-service.rst [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+``CorsServiceBuilder.forOrigin``, e.g.
+
+.. code-block:: java
+
+    HttpService myService = new AbstractHttpService() {
```
Comment (trustin) [inline site/src/sphinx/server-cors-service.rst]: `myService = (ctx, req) -> ...;

File: site/src/sphinx/server-cors-service.rst [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+                              .allowNullOrigin() // this property will allow "null" origin
+                              .allowRequestMethods(HttpMethod.POST, HttpMethod.GET)
+                              .allowRequestHeaders(HttpHeaderNames.of("allow_request_header"))
+                              .exposeHeaders(HttpHeaderNames.of("expose_header_1"),
+                                             HttpHeaderNames.of("expose_header_2"))
+                              .preflightResponseHeader("x-preflight-cors", "Hello CORS")
+                              .newDecorator()));
+
+
+Applying a different policy for different origins
```
Comment (trustin) [inline site/src/sphinx/server-cors-service.rst]: different policies for different origins?

File: site/src/sphinx/server-cors-service.rst [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+-------------------------------------------------
+To configure multiple policies for different groups of origins, use ``andForOrigins`` or ``andForOrigin`` method
```
Comment (trustin) [inline site/src/sphinx/server-cors-service.rst]: - multiple -> different
- andForOrigins or andForOrigin -> andForOrigins() or andForOrigin()

File: site/src/sphinx/server-cors-service.rst [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+which creates a new instance of :api:`ChainedCorsPolicyBuilder` with the specified origins and
```
Comment (trustin) [inline site/src/sphinx/server-cors-service.rst]: - `s/instance of //`
- and you have to: How about ending the sentence right after 'the specified origins'? i.e.

   ... with the specified origins. Call ``and()`` to return to CorsServiceBuilder 
   once you are done with building a policy, e.g.

File: site/src/sphinx/server-cors.rst [hunk 10]
```diff
@@ -0,0 +1,107 @@
+.. _server-cors-service:
+
+Configure CORS
+========================
+
+Armeria provides a way to configure Cross-origin resource sharing (CORS) policy for specified origins or
+any origin via :api:`CorsServiceBuilder`. For more information about CORS,
+Visit the Wikipedia's CORS <https://en.wikipedia.org/wiki/Cross-origin_resource_sharing>.
```
Comment (trustin) [inline site/src/sphinx/server-cors.rst]: You need to use a backtick to create a link:

```rst
Visit `Wikipedia's CORS page <https://.....>`_.
```

File: site/src/sphinx/server-cors.rst [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+
+Allowing any origin
+-------------------
+To configure CORS Service allowing any origin (*), use ``CorsServiceBuilder.forAnyOrigin()``
+
+.. code-block:: java
+
+    HttpService myService = (ctx, req) -> {
+                ...
+    }
```
Comment (trustin) [inline site/src/sphinx/server-cors.rst]: No need to keep `{` and `}`? `(ctx, req) -> ...;` would save more space.

File: site/src/sphinx/server-cors.rst [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+Visit the Wikipedia's CORS <https://en.wikipedia.org/wiki/Cross-origin_resource_sharing>.
+
+
+Allowing any origin
+-------------------
+To configure CORS Service allowing any origin (*), use ``CorsServiceBuilder.forAnyOrigin()``
+
+.. code-block:: java
+
+    HttpService myService = (ctx, req) -> {
```
Comment (trustin) [inline site/src/sphinx/server-cors.rst]: How about adding imports for the type that appears first time in this page?

File: site/src/sphinx/server-cors.rst [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+
+
+Allowing specified origins
+------------------------------
+To configure CORS Service allowing specified origins, use ``CorsServiceBuilder.forOrigins()`` or
+``CorsServiceBuilder.forOrigin()``, e.g.
+
+.. code-block:: java
+
+    HttpService myService = (ctx, req) -> {
```
Comment (trustin) [inline site/src/sphinx/server-cors.rst]: `(ctx, req) -> ...;`

File: site/src/sphinx/server-cors.rst [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+            CorsServiceBuilder.forOrigins("http://example.com")
+                              .allowCredential()
+                              .allowNullOrigin() // this property will allow "null" origin
+                              .allowRequestMethods(HttpMethod.POST, HttpMethod.GET)
+                              .allowRequestHeaders("allow_request_header")
+                              .exposeHeaders("expose_header_1", "expose_header_2")
+                              .preflightResponseHeader("x-preflight-cors", "Hello CORS")
+                              .maxAge(3600)
+                              .and()
+                              .addPolicy(new CorsPolicyBuilder("http://example2.com")
```
Comment (trustin) [inline site/src/sphinx/server-cors.rst]: Please add import.

File: site/src/sphinx/server-cors.rst [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+                              .allowRequestHeaders("allow_request_header2")
+                              .exposeHeaders("expose_header_3", "expose_header_4")
+                              .and()
+                              .newDecorator()));
+
+You can also directly add a :api:`CorsPolicy` created by a :api:`CorsPolicyBuilder`, e.g.
+
+.. code-block:: java
+
+    HttpService myService = (ctx, req) -> {
```
Comment (trustin) [inline site/src/sphinx/server-cors.rst]: { ... }; -> ...;

File: site/src/sphinx/server-cors.rst [hunk 16]
```diff
@@ -0,0 +1,107 @@
+.. _server-cors-service:
```
Comment (trustin) [inline site/src/sphinx/server-cors.rst]: `.. _server-cors:`

File: site/src/sphinx/server.rst [hunk 17]
```diff
@@ -15,3 +15,4 @@ Writing a server
     server-http-file
     server-servlet
     server-access-log
+    server-cors-service
```
Comment (trustin) [inline site/src/sphinx/server.rst]: `server-cors`

File: site/src/sphinx/server-cors.rst [hunk 18]
```diff
@@ -0,0 +1,106 @@
+.. _server-cors:
+
+Configure CORS
+========================
+
+Armeria provides a way to configure Cross-origin resource sharing (CORS) policy for specific origins or
+any origin via :api:`CorsServiceBuilder`. For more information about CORS,
+Visit `Wikipedia's CORS <https://en.wikipedia.org/wiki/Cross-origin_resource_sharing>`_.
```
Comment (minwoox) [inline site/src/sphinx/server-cors.rst]: visit? because it is not the first letter of this sentence?

File: site/src/sphinx/server-cors.rst [hunk 19]
```diff
@@ -0,0 +1,106 @@
+.. _server-cors:
+
+Configure CORS
```
Comment (minwoox) [inline site/src/sphinx/server-cors.rst]: How about matching the tense of this subhead with the subheads below?
`Configuring CORS` or `Allow any...`

File: site/src/sphinx/server-cors.rst [hunk 20]
```diff
@@ -1,11 +1,11 @@
...
+.. _server-cors:
 
 Configure CORS
 ========================
 
-Armeria provides a way to configure Cross-origin resource sharing (CORS) policy for specified origins or
+Armeria provides a way to configure Cross-origin resource sharing (CORS) policy for specific origins or
 any origin via :api:`CorsServiceBuilder`. For more information about CORS,
-Visit the Wikipedia's CORS <https://en.wikipedia.org/wiki/Cross-origin_resource_sharing>.
+Visit `Wikipedia's CORS <https://en.wikipedia.org/wiki/Cross-origin_resource_sharing>`_.
```
Comment (trustin) [inline site/src/sphinx/server-cors.rst]: CORS -> CORS page

File: site/src/sphinx/server-cors.rst [hunk 21]
```diff
@@ -0,0 +1,106 @@
...
+            CorsServiceBuilder.forAnyOrigin()
+                              .allowCredentials()
+                              .allowRequestMethods(HttpMethod.POST, HttpMethod.GET)
+                              .allowRequestHeaders("allow_request_header")
+                              .exposeHeaders("expose_header_1", "expose_header_2")
+                              .preflightResponseHeader("x-preflight-cors", "Hello CORS")
+                              .newDecorator()));
+
+
+
```
Comment (minwoox) [inline site/src/sphinx/server-cors.rst]: triple empty lines.

File: site/src/sphinx/server-cors.rst (hunk 19 above)
Comment (tmdgus0118) [reply site/src/sphinx/server-cors.rst]: Changed to `Configuring`! Thanks.

**Comment by minwoox** [comment]
Do we need to remove the CORS part in the gRPC service documentation? WDYT @trustin?

**Comment by minwoox** [comment]
Or, just update the link?

**Comment by tmdgus0118** [comment]
+1 for just updating the link.

File: site/src/sphinx/server-cors.rst [hunk 22]
```diff
@@ -0,0 +1,104 @@
+.. _server-cors:
+
+Configuring CORS
+========================
```
Comment (hyangtack) [inline site/src/sphinx/server-cors.rst]: Please shorten `=` as the same length of the title.

File: site/src/sphinx/server-cors.rst [hunk 23]
```diff
@@ -0,0 +1,104 @@
...
+            CorsServiceBuilder.forAnyOrigin()
+                              .allowCredentials()
+                              .allowRequestMethods(HttpMethod.POST, HttpMethod.GET)
+                              .allowRequestHeaders("allow_request_header")
+                              .exposeHeaders("expose_header_1", "expose_header_2")
+                              .preflightResponseHeader("x-preflight-cors", "Hello CORS")
+                              .newDecorator()));
+
+Allowing specific origins
+------------------------------
```
Comment (hyangtack) [inline site/src/sphinx/server-cors.rst]: Ditto

File: site/src/sphinx/server-cors.rst [hunk 24]
```diff
@@ -0,0 +1,104 @@
...
+========================
+
+Armeria provides a way to configure Cross-origin resource sharing (CORS) policy for specific origins or
+any origin via :api:`CorsServiceBuilder`. For more information about CORS,
+visit `Wikipedia's CORS page <https://en.wikipedia.org/wiki/Cross-origin_resource_sharing>`_.
+
+
+Allowing any origin
+-------------------
+To configure CORS Service allowing any origin (*), use ``CorsServiceBuilder.forAnyOrigin()``
```
Comment (hyangtack) [inline site/src/sphinx/server-cors.rst]: Is it better to add `, e.g.` like below?

File: site/src/sphinx/server-cors.rst [hunk 25]
```diff
@@ -0,0 +1,104 @@
...
+To configure CORS Service allowing specific origins, use ``CorsServiceBuilder.forOrigins()`` or
+``CorsServiceBuilder.forOrigin()``, e.g.
+
+.. code-block:: java
+
+    HttpService myService = (ctx, req) -> ...;
+    ServerBuilder sb = new ServerBuilder().service("/message", myService.decorate(
+            CorsServiceBuilder.forOrigins("http://example.com")
+                              .allowCredentials()
+                              .allowNullOrigin() // this property will allow "null" origin
```
Comment (hyangtack) [inline site/src/sphinx/server-cors.rst]: How about `'Origin: null' will be accepted.`?

**Comment by trustin** [comment]
+1 for just updating the link. @tmdgus0118 Could you address it in this PR?

**Comment by trustin** [comment]
Good job, @tmdgus0118 :-)

**Pull request merged**